resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -1464,6 +1464,34 @@ private object Client extends Logging {
     (mainUri ++ secondaryUris).toArray
   }
 
+  /**
+   * Returns a list of local, absolute file URLs representing the user classpath. Note that this
+   * must be executed on the same host which will access the URLs, as it will resolve relative
+   * paths based on the current working directory.
+   *
+   * @param conf Spark configuration.
+   * @param useClusterPath Whether to use the 'cluster' path when resolving paths with the
+   *                       `local` scheme. This should be used when running on the cluster, but
+   *                       not when running on the gateway (i.e. for the driver in `client` mode).
+   * @return Array of local URLs ready to be passed to a [[java.net.URLClassLoader]].
+   */
+  def getUserClasspathUrls(conf: SparkConf, useClusterPath: Boolean): Array[URL] = {
+    Client.getUserClasspath(conf).map { uri =>
+      val inputPath = uri.getPath
+      val localPath = if (Utils.isLocalUri(uri.toString)) {
```

Comment:
so this isn't exactly the same as the logic in ExecutorRunnable since this only applies getClusterPath if they have local: scheme and it applied to any absolute path.  I would have expected that the getClusterPath applied to just file: as well.   is there a reason we didn't just keep it if absolute path?
Really if someone specifies local: they mean it will be at that location on all the worker nodes in that location, I don't see that the original issue that added this said anything about local: (https://issues.apache.org/jira/browse/SPARK-8302)

Reply:
there is a test in ClientSuite.scala  but it doesn't cover this part, it just covers client side, there it tests that absolute path without local: is changed.

Reply:
Hm, thanks for the context here. This is a good question.

This is specifically for the classpath, which is populated via `--jars`, `spark.yarn.dist.jars`, or `spark.jars`. For any of these options, if you pass a raw path (e.g. `/path/foo.jar`) or a `file:` path (e.g. `file:/path/foo.jar`), those will get interpreted locally on the client side and uploaded into the distributed cache, so by the time we reach `getUserClasspathUrls`, we're using the localized path, not the original path -- except in the case of `local:`, which explicitly bypasses the distributed cache:
https://github.com/apache/spark/blob/e8d6992b664cf668ed5ad12f891b2341f0095c55/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala#L491

Since these non-`local:` paths are set by Spark/YARN instead of the user, they will already point to the correct location on the cluster machines, and shouldn't go through replacement.

Regarding the `Cluster path translation` test in `ClientSuite`, the part not using `local:` is testing `extraClassPath`, which is a list of classpath entries vs. a list of URIs, and doesn't get put into the distributed cache, so it's correct to perform replacement there even for non-`local:` URIs.

Let me know if you think I'm missing anything!

Reply:
ah yes you are right, this is only looking at the jars we uploaded, was thinking of other class path, but those are handled differently.